File: iOS/HandGestureRecognizing/HandGestureRecognizing/HandGestureRecognizing/PrefixMatcher.swift
import Foundation
import GestureModelModule

/// Pure Phase 2 prefix-matching state machine.
///
/// No threading, no timers — feed pose predictions one at a time (on each detected
/// hold), read the returned Action. HandGestureRecognizing owns the T_commit timer
/// and calls `commitCurrentMatch()` when it fires.
///
/// Implements plan §Phase 2 Runtime flow rules for regular, idle, and
/// unconfirmed poses.
public final class PrefixMatcher {

    // MARK: - Action

    public enum Action {
        /// No template starts with `observed` — discard and reset the gate.
        case noPrefix
        /// `observed` is a live prefix only; no complete match yet — keep buffering.
        case livePrefix
        /// Complete match, no longer prefix possible — commit with `candidateSet`.
        case commitNow(candidateSet: Set<String>)
        /// Complete match exists but longer prefixes are still possible —
        /// start T_commit; commit `candidateSet` if it fires.
        case startCommitTimer(candidateSet: Set<String>)
        /// Idle pose on empty `observed` — reset gate, keep watching.
        case idleReset
        /// Idle pose with no complete match and no complete ancestor — discard.
        case idleDiscard(candidateSet: Set<String>)
        /// Idle pose on complete match — commit immediately.
        case idleCommit(candidateSet: Set<String>)
    }

    // MARK: - State

    private let manifest: PoseManifest
    public private(set) var observedSequence: [Int] = []

    public init(manifest: PoseManifest) {
        self.manifest = manifest
    }

    // MARK: - Public API

    public func reset() {
        observedSequence.removeAll()
    }

    /// Returns the candidate set if `observed` has any complete template match,
    /// for use on gate-close (the gate-close commit path).
    public func gateCloseCommitSet() -> Set<String>? {
        let matches = gesturesMatchingExactly(observedSequence)
        return matches.isEmpty ? nil : matches
    }

    /// Feed one pose prediction and get back the action HandGestureRecognizing should take.
    public func observe(poseId: Int, kind: ClusterKind) -> Action {
        switch kind {
        case .unconfirmed:
            // Unconfirmed clusters are rejected at runtime — same as sub-τ confidence.
            return .noPrefix

        case .idle:
            return handleIdle()

        case .regular:
            return handleRegular(poseId: poseId)
        }
    }

    // MARK: - Private — idle path

    private func handleIdle() -> Action {
        if observedSequence.isEmpty {
            return .idleReset
        }
        let complete = gesturesMatchingExactly(observedSequence)
        if !complete.isEmpty {
            return .idleCommit(candidateSet: complete)
        }
        // Live prefix with no complete match — find longest complete ancestor
        if let ancestor = longestCompleteAncestor() {
            return .idleCommit(candidateSet: ancestor)
        }
        return .idleDiscard(candidateSet: Set(manifest.gestureTemplates.keys))
    }

    // MARK: - Private — regular path

    private func handleRegular(poseId: Int) -> Action {
        let candidate = observedSequence + [poseId]

        // Does any template (across all gestures) start with this candidate?
        let gesturesWithPrefix = manifest.gestureTemplates.filter { (_, templates) in
            templates.contains { template in
                template.count >= candidate.count &&
                Array(template.prefix(candidate.count)) == candidate
            }
        }

        guard !gesturesWithPrefix.isEmpty else {
            return .noPrefix
        }

        // Commit candidate as the new observed
        observedSequence = candidate

        // Which gestures have an exact template match?
        let exactMatches = gesturesMatchingExactly(candidate)
        let longerPossible = gesturesWithPrefix.values.contains { templates in
            templates.contains { $0.count > candidate.count }
        }

        if exactMatches.isEmpty {
            // Only live prefixes — keep waiting
            return .livePrefix
        } else if !longerPossible {
            // Complete match, no longer prefix possible — commit immediately
            return .commitNow(candidateSet: exactMatches)
        } else {
            // Complete match but longer still possible — start T_commit timer
            return .startCommitTimer(candidateSet: exactMatches)
        }
    }

    // MARK: - Private helpers

    private func gesturesMatchingExactly(_ sequence: [Int]) -> Set<String> {
        Set(manifest.gestureTemplates.compactMap { (gestureId, templates) in
            templates.contains { $0 == sequence } ? gestureId : nil
        })
    }

    /// Longest strict prefix of `observedSequence` that is a complete template match.
    private func longestCompleteAncestor() -> Set<String>? {
        for length in stride(from: observedSequence.count - 1, through: 1, by: -1) {
            let prefix = Array(observedSequence.prefix(length))
            let matches = gesturesMatchingExactly(prefix)
            if !matches.isEmpty { return matches }
        }
        return nil
    }

    // MARK: - Commit

    /// Called by HandGestureRecognizing when the T_commit timer fires.
    /// Returns the candidate set that was pending at timer start — the caller
    /// is responsible for not calling this if the gate has already closed.
    public func commitTimerFiredSet(for candidateSet: Set<String>) -> Set<String> {
        // Re-validate: gestures in the set that still have a template prefixed by observed
        let valid = candidateSet.filter { gestureId in
            guard let templates = manifest.gestureTemplates[gestureId] else { return false }
            return templates.contains { template in
                Array(template.prefix(observedSequence.count)) == observedSequence
            }
        }
        return valid.isEmpty ? candidateSet : valid
    }
}

File: core/src/hand_gesture_recognizing/PrefixMatcher.cpp
#include "PrefixMatcher.hpp"
#include <algorithm>
#include <map>

PrefixMatcher::PrefixMatcher(const CgPoseManifest& manifest) : manifest_(manifest) {}

void PrefixMatcher::reset() {
    observed_.clear();
}

std::optional<std::set<std::string>> PrefixMatcher::gateCloseCommitSet() const {
    auto matches = gesturesMatchingExactly(observed_);
    if (matches.empty()) return std::nullopt;
    return matches;
}

PrefixMatcher::Action PrefixMatcher::observe(int pose_id, CgClusterKind kind) {
    switch (kind) {
    case CgClusterKind::unconfirmed:
        return {Action::Kind::no_prefix};
    case CgClusterKind::idle:
        return handleIdle();
    case CgClusterKind::regular:
        return handleRegular(pose_id);
    }
    return {Action::Kind::no_prefix};
}

std::set<std::string> PrefixMatcher::commitTimerFiredSet(
        const std::set<std::string>& candidate_set) const {
    std::set<std::string> valid;
    for (const auto& gesture_id : candidate_set) {
        auto it = manifest_.gesture_templates.find(gesture_id);
        if (it == manifest_.gesture_templates.end()) continue;
        for (const auto& tmpl : it->second) {
            if (tmpl.size() >= observed_.size()) {
                auto prefix = std::vector<int>(tmpl.begin(),
                                               tmpl.begin() + observed_.size());
                if (prefix == observed_) { valid.insert(gesture_id); break; }
            }
        }
    }
    return valid.empty() ? candidate_set : valid;
}

// --------------------------------------------------------------------------
// Private
// --------------------------------------------------------------------------

PrefixMatcher::Action PrefixMatcher::handleIdle() {
    if (observed_.empty()) return {Action::Kind::idle_reset};

    auto complete = gesturesMatchingExactly(observed_);
    if (!complete.empty()) return {Action::Kind::idle_commit, complete};

    if (auto ancestor = longestCompleteAncestor()) {
        return {Action::Kind::idle_commit, *ancestor};
    }

    // Build full gesture ID set as the discard set.
    std::set<std::string> all;
    for (const auto& [id, _] : manifest_.gesture_templates) all.insert(id);
    return {Action::Kind::idle_discard, all};
}

PrefixMatcher::Action PrefixMatcher::handleRegular(int pose_id) {
    std::vector<int> candidate = observed_;
    candidate.push_back(pose_id);

    // Does any template start with `candidate`?
    std::map<std::string, bool> gestures_with_prefix; // id → has_longer
    for (const auto& [gesture_id, templates] : manifest_.gesture_templates) {
        bool any_prefix = false, any_longer = false;
        for (const auto& tmpl : templates) {
            if (tmpl.size() < candidate.size()) continue;
            auto prefix = std::vector<int>(tmpl.begin(),
                                           tmpl.begin() + candidate.size());
            if (prefix == candidate) {
                any_prefix = true;
                if (tmpl.size() > candidate.size()) any_longer = true;
            }
        }
        if (any_prefix) gestures_with_prefix[gesture_id] = any_longer;
    }

    if (gestures_with_prefix.empty()) return {Action::Kind::no_prefix};

    observed_ = candidate;

    auto exact = gesturesMatchingExactly(candidate);
    bool longer_possible = false;
    for (const auto& [_, has_longer] : gestures_with_prefix) {
        if (has_longer) { longer_possible = true; break; }
    }

    if (exact.empty()) return {Action::Kind::live_prefix};
    if (!longer_possible) return {Action::Kind::commit_now, exact};
    return {Action::Kind::start_commit_timer, exact};
}

std::set<std::string> PrefixMatcher::gesturesMatchingExactly(
        const std::vector<int>& seq) const {
    std::set<std::string> result;
    for (const auto& [gesture_id, templates] : manifest_.gesture_templates) {
        for (const auto& tmpl : templates) {
            if (tmpl == seq) { result.insert(gesture_id); break; }
        }
    }
    return result;
}

std::optional<std::set<std::string>> PrefixMatcher::longestCompleteAncestor() const {
    for (int len = static_cast<int>(observed_.size()) - 1; len >= 1; --len) {
        auto prefix = std::vector<int>(observed_.begin(), observed_.begin() + len);
        auto matches = gesturesMatchingExactly(prefix);
        if (!matches.empty()) return matches;
    }
    return std::nullopt;
}
